### PR TITLE
feat: add `Table` && `TableTestAccessor` && table utilities

### PR DESCRIPTION
**Pull request by iajoiner**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [x] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [x] I have run the ci check script with `source scripts/run_ci_checks.sh`.

# Rationale for this change
In order to make `ProofPlan`s composable it is helpful to add `Table` which will replace `DataAccessor` in `ProofExpr`. In order to add tests that aren't too tedious it is necessary to add table utilities analogous to owned table ones.

# What changes are included in this PR?
- add `Table`
- add table utilities in `table_utility.rs`
- add `TableTestAccessor

# Are these changes tested?

Yes.

### CODE REVIEW

File: crates/proof-of-sql/src/base/database/table_test.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
+use crate::{
+    base::{
+        database::{table_utility::*, Column, Table, TableError},
+        map::IndexMap,
+        scalar::test_scalar::TestScalar,
+    },
+    proof_primitive::dory::DoryScalar,
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/table_test.rs:7]: This causes a cyclic dependency.
Nothing in `base` should depend on anything outside of `base`.

File: crates/proof-of-sql/src/base/database/table_utility.rs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+/// use proof_of_sql::base::{database::table_utility::*, scalar::Curve25519Scalar};
+/// let alloc = Bump::new();
+/// let result = table::<Curve25519Scalar>([
+///     borrowed_bigint("a", [1, 2, 3], &alloc),
+/// ]);
+/// ```
+///
+/// # Panics
+/// - Panics if `name.parse()` fails to convert the name into an `Identifier`.
+#[allow(clippy::missing_panics_doc)]
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/table_utility.rs:144]:
```suggestion
```

File: crates/proof-of-sql/src/base/database/table_test_accessor.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/table_test_accessor.rs:1]: Is this intended to be a replacement for `OwnedTableTestAccessor`?

Comment (iajoiner) [reply crates/proof-of-sql/src/base/database/table_test_accessor.rs:1]: @JayWhite2357 It's just something fairly easy to work with when one needs an accessor. Yes it is more straightforward than `OwnedTableTestAccessor` but I don't plan to do a full replacement now.

File: crates/proof-of-sql/src/base/database/table.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    /// Number of columns in the table.
+    #[must_use]
+    pub fn num_columns(&self) -> usize {
+        self.table.len()
+    }
+    /// Number of rows in the table.
+    #[must_use]
+    pub fn num_rows(&self) -> usize {
+        if self.table.is_empty() {
+            0
```
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/base/database/table.rs:50]: @JayWhite2357 Good idea. Or sometimes unwrap_or(1).

**Review by JayWhite2357 — APPROVED** [review]

**Pull request merged**